src/lib.rs

```diff
@@ -954,6 +954,22 @@ pub mod sync {
                 None => panic!("Lazy instance has previously been poisoned"),
             })
         }
+
+        /// Forces the evaluation of this lazy value and
+        /// returns the result.
+        ///
+        /// # Example
+        /// ```
+        /// use once_cell::sync::Lazy;
+        ///
+        /// let lazy = Lazy::new(|| 92);
+        ///
+        /// assert_eq!(lazy.into_inner(), 92);
+        /// ```
+        pub fn into_inner(self) -> T {
+            Lazy::force(&self);
+            self.cell.into_inner().unwrap()
+        }
```

Comment:
I feel that forcing and destructing should be orthogonal, so I don't feel good about a function that does both. My gut feeling would be that the signature should look like `fn into_inner(self) -> Option<T>`, but then we loose the `F`...

Reply:
Returning `Option<T>` will give away the information of whether the callback has been called or not, which is something that the current `Lazy` interface does not provide. I kind of like it that way, which is why I chose to `force` the value before returning it. If people feel strongly about introspection, maybe they should use `OnceCell` instead of `Lazy`?

Reply:
I see now problem adding `Lazy::is_empty(this: &Self) -> bool` method, it's just that nobody asked for that. 

In general, I feel that APIs for data-strcuture things should not be prescriptive, and should err on the side of generality. Basically, let the user decide how to use the thing.

Reply:
What is the plan of moving forward?

If we introduce a destructor with type `Lazy<T> -> Result<T, FnOnce() -> T>`, it likely should go under a different name than `into_inner`, which by convention should have type `Lazy<T> -> T`. Maybe call it `extract`?

In any case, I'm not opposed to adding `extract` or `is_empty` to the interface, and I don't think they are mutually exclusive to an `into_inner` function either. Maybe they can go into a separate PR. What do you think? @matklad

Reply:
How about naming this method `into_value`? It seems useful as is, just not appropriately named.

Reply:
Hm, I still think that neither name nor signature are optimal. Let's dodge the issue by introducing a roughly equivalent API, which does have a perfect name:

```
impl Lazy<T> {
  fn take(&mut self) -> Option<T> { ... }
}
```

Reply:
I'm surprised that `take` was added to `OnceCell`. I thought a central tenant of `OnceCell` is "can be assigned to at most once" (at the top of the readme). Isn't that the meaning of the name of the struct? Adding `take` to `Lazy` would remove the (I think important) semantic that the initializer will run at most once.

A positive of having a function that does both forcing and destructing is that the synchronization can (at least potentially) be optimized as a single operation. To offer one more name idea: `force_into_value`.

Reply:
(I'm the one that added take.)

once_cell's at-most-once guarantee applies in the shared reference context. If you have a mutable reference you can do whatever with it. (I had a comment explaining this but @matklad removed it I think ;) )

Reply:
That makes sense.

Reply:
I think the functionality of `into_inner` here is not merely a convenience. Yes, now with `take` we could implement a similar thing, but it is awkward to have an unnecessary unwrap in there.

Reply:
Right. Perhaps there could be `force_take(&mut self) -> T`. I don't mean to flippantly keep adding ideas. I'm hoping something will stick that we can agree on.